### PR TITLE
Foundation API Conformance Update: Errors

### PR DESCRIPTION
**Pull request by itaiferber**
### Changes
* `NSCocoaError` → `CocoaError`
* Reimplement top-level error values in terms of consistent definitions
* Import new error types (`RecoverableError`, `LocalizedError`) from overlay
* Import URL errors and POSIX errors from overlay
* Import error bridging from overlay
* Update error code usage throughout project

### CODE REVIEW

**Comment by itaiferber** [comment]
@swift-ci Please test

**Comment by itaiferber** [comment]
@DougGregor Any thoughts/concerns about this change? I ported the `NSError` changes almost wholesale from the overlay (sans features requiring silgen and forms of bridging which are not present on Linux).

**Comment by itaiferber** [comment]
Hmm. Looks like a serious compiler crash. Can't replicate on Linux myself, but looking into it.

**Comment by itaiferber** [comment]
Perhaps there was a timing issue involved here. I want to see if re-testing makes any sort of difference. @swift-ci Please test

**Comment by itaiferber** [comment]
(For reference, the failure is still available to view [here](https://ci.swift.org/job/swift-corelibs-foundation-PR-Linux/241/console) while CI re-runs)

File: Foundation/NSError.swift [hunk 1]
```diff
@@ -187,29 +196,218 @@ extension CFError : _NSBridgable {
...
+    return result
+}
+
+// NSError and CFError conform to the standard Error protocol. Compiler
+// magic allows this to be done as a "toll-free" conversion when an NSError
+// or CFError is used as an Error existential.
+
+extension NSError {
+    /// The "embedded" NSError is itself.
+    public func _getEmbeddedNSError() -> AnyObject? {
```
Comment (DougGregor) [inline Foundation/NSError.swift]: `_getEmbeddedNSError` isn't actually used when we don't have an Objective-C runtime, so you can remove this bit.

File: Foundation/NSError.swift [hunk 2]
```diff
@@ -187,29 +196,218 @@ extension CFError : _NSBridgable {
...
+    public var _code: Int {
+        return CFErrorGetCode(self)
+    }
+
+    public var _userInfo: Any? {
+        return CFErrorCopyUserInfo(self) as Any
+    }
+
+    /// The "embedded" NSError is itself.
+    public func _getEmbeddedNSError() -> AnyObject? {
```
Comment (DougGregor) [inline Foundation/NSError.swift]: Same here

File: Foundation/NSError.swift [hunk 3]
```diff
@@ -187,29 +196,218 @@ extension CFError : _NSBridgable {
...
+        return self
+    }
+}
+
+// An error value to use when an Objective-C API indicates error
+// but produces a nil error object.
+public enum _GenericObjCError : Error {
```
Comment (DougGregor) [inline Foundation/NSError.swift]: This is only used when bridging NSError, so I don't think you need this in the corelibs Foundation.

File: Foundation/NSError.swift [hunk 4]
```diff
@@ -244,7 +445,909 @@ public extension __BridgedNSError where RawValue: UnsignedInteger {
...
+            return nil
+        }
+
+        self.init(_nsError: error)
+    }
+}
+
+/// Implementation of CustomNSError for all _BridgedStoredNSErrors.
+public extension _BridgedStoredNSError {
+    // FIXME: Would prefer to have a clear "extract an NSError
```
Comment (DougGregor) [inline Foundation/NSError.swift]: Oops, we should kill this FIXME in the overlay; this is what `getEmbeddedNSError` does.

File: Foundation/NSError.swift (hunk 1 above)
Comment (itaiferber) [inline Foundation/NSError.swift]: Figured, but wasn't 100% sure (about this and about `_GenericObjCError`); will remove.

**Comment by itaiferber** [comment]
I doubt this will fix the CI build error, but it's worth a shot. @swift-ci Please test

**Comment by itaiferber** [comment]
@swift-ci Please test

**Comment by DougGregor** [comment]
The compiler crash looks like we're trying to do some bridging without checking for ObjC interop.

**Comment by DougGregor** [comment]
For reference, the rest of this PR looks good to me. Thank you!

**Comment by itaiferber** [comment]
@DougGregor Figured; trying to diagnose, though haven't been able to reproduce on Linux yet, which will make testing expensive if I have to push to CI every time.

Thanks for the help!

**Comment by itaiferber** [comment]
Pushed a change that might avoid the compilation crash we see here.

**Comment by itaiferber** [comment]
@swift-ci Please test

**Comment by itaiferber** [comment]
Looks like that fixed the `hashValue` failure point; the only other location is `isEqual`. Hopefully this change will correct things. @swift-ci Please test

**Comment by itaiferber** [comment]
@swift-ci Please test

**Comment by itaiferber** [comment]
Perfect. Avoiding the (missing) bridging by overriding `NSObject` methods avoids the silgen issue.

**Pull request merged**